src/lib/y2partitioner/widgets/format_and_mount.rb

```diff
@@ -612,6 +612,17 @@ def refresh
         self.value = @controller.encrypt
       end
 
+      # validates if it can be encrypted
+      def validate
+        return true unless value
+
+        # size number is from bsc1065071#c5
+        return true if @controller.blk_device.size > ::Y2Storage::DiskSize.MiB(2)
+
+        ::Yast::Popup.Error(_("For encryption partition have to be bigger then 2 MiB."))
+        false
+      end
```

Comment:
I see a problem with this approach. The minimum size depends on several aspects, not only on the fact the device is formatted. For example, different filesystem types could require different min sizes.

A part of that, there are more checks that have to be done an the end of this screen. For example, we have to check that the user gives a filesystem label when the default mount_by is set to label, see [my PR](https://github.com/yast/yast-storage-ng/pull/529/commits/435abbc51a204d4af9da85aa831f3ac9fc72014f).

Maybe it is time to create a class (e.g. `Controllers::Filesystem::Validations` or something similar) to manage all validations needed in that screen, reporting the list of errors.

Reply:
well, goal of this is to check real minimal for optimistic approach. I do not opposite that it can be improved, but I need to know how to improve.

Reply:
In theory, `blk_device.resize_info` can be used to know the minimum size based on everything the block device contains (filesystem of a given type, encryption and so on). If we go for the "cheap" solution (independent validations placed in random widgets with hardcoded limits), I would like to see a big TODO/FIXME next to each one. Explaining that it should be centralized and mentioning `blk_device.resize_info`.

Reply:
@ancorgs well, problem with resize_info is that I have to first create that encryption device to know it, but I worry it is too late. Of course we can add generic checkk if resize_info for each partition is bigger then size, but for this one I want quick check in widget to prevent continue.
So do you think it makes sense to also add generic check that check if all partition have size at least as big as minimal in resize_info?

Reply:
Of course, we may need to change a little bit how some things are done to achieve that. But in the mid term I want **only one** check for the size of the device relying on the library. For the future, I don't want this check (with a narrow-scoped limit just hard-coded somewhere in a widget) and **additionally** the one based on `resize_info`.

In the future, the current approach can lead to a warning saying the 2MiB is too small (only local widget check). Then the user changes it to 4MiB and it gets a second warning coming from the filesystem_type widget saying that XFS requires at least 5 MiB (I just made up the number). Then he changes it to 6MiB and that's still wrong because the combination of 2 and 5 means the real minimum is 7MiB.

_Beware: the rest is actually another more broad topic. I just couldn't resist_

That problem illustrate want of the things I dislike about the CWM "default philosophy" of having domain-related logic (like validation and similar stuff) spread over a collection of widgets and subwidgets. In the partitioner we are trying to limit that by placing as much domain-related logic as possible into the classes within the `Actions` namespace. So we have only one place where we validate the different aspects of the device being created, instead of having that scattered over a dozen of CWM classes. So our widgets are as dumb as possible, instead of being super-smart components.

Reply:
Just to clarify. For the time being, I'm just asking for a FIXME/TODO mentioning a more centralized and library-based solution for the future. Not to change anything right now.

Reply:
@ancorgs to comment last paragraph. CWM validation is useful when you have invalid value that is local to one widget. What you want is probably due to complexity check that verify whole partition setup after whole wizard and this cannot be done on CWM level and more probably on sequencer level which checks it as last step and allow going back is something is wrong.

Reply:
Yes, in this case the reason is that the check is complex and implies several aspects. But even in simpler cases, I believe that having the validations about a given device in a single place makes more sense than a lot of local "validate" methods scattered all around in widgets.

So is not "we have to centralize here because is a complex check". For me is more like "we have to centralize because is the sane approach, period" :wink:

Anyway, we are getting too much out of scope. My fault. :wink: Let's leave this for some beer-powered discussion.